packages/grpc-js/src/server-call.ts

```diff
@@ -439,6 +442,56 @@ export class Http2ServerCallStream<
     return this.cancelled;
   }
 
+  private getDecompressedMessage(message: Buffer, encoding: string) {
+    switch (encoding) {
+      case 'deflate': {
+        return new Promise<Buffer | undefined>((resolve, reject) => {
+          zlib.inflate(message.slice(5), (err, output) => {
+            if (err) {
+              this.sendError({
+                code: Status.INTERNAL,
+                details: `Received "grpc-encoding" header "${encoding}" but ${encoding} decompression failed`,
+              });
+              resolve();
+            } else {
+              const joined = Buffer.concat([message.slice(0, 5), output]);
+              resolve(joined);
+            }
+          });
+        });
+      }
+  
+      case 'gzip': {
+        return new Promise<Buffer | undefined>((resolve, reject) => {
+          zlib.unzip(message.slice(5), (err, output) => {
+            if (err) {
+              this.sendError({
+                code: Status.INTERNAL,
+                details: `Received "grpc-encoding" header "${encoding}" but ${encoding} decompression failed`,
+              });
+              resolve();
+            } else {
+              const joined = Buffer.concat([message.slice(0, 5), output]);
+              resolve(joined);
+            }
+          });
+        });
+      }
+
+      case 'identity': {
+        return Promise.resolve(message);
+      }
+  
+      default: {
+        this.sendError({
+          code: Status.UNIMPLEMENTED,
+          details: `Received message compressed with unsupported encoding "${encoding}"`,
+        });
+        return Promise.resolve();
+      }
+    }
+  }
```

Comment:
I didn't notice this before, but this function has to actually remove the 5-byte chunk at the beginning of the message. That chunk contains the compressed flag and the length of the encoded message, which are both invalid and useless after the message has been decompressed. Then `deserializeMessage` also has to be modified to not slice off the first 5 bytes.

Reply:
@murgatroid99 It does remove the 5-byte chunk before decompressing it; do you mean to avoid re-attaching it before returning it?
```
              const joined = Buffer.concat([message.slice(0, 5), output]);
              resolve(joined);
```

Which would obviate the need to remove it again later:
```
  deserializeMessage(bytes: Buffer) {
    // TODO(cjihrig): Call compression aware deserializeMessage().
    const receivedMessage = bytes.slice(5);

    return this.handler.deserialize(receivedMessage);
  }
```

If that's what you meant, I'll make that change.

On a separate note, I'm not able to reproduce the test failures in `Linux Tests`.  When I run the tests locally (on Ubuntu 20.04.1 via WSL2), the following two tests fail:
```
  1) Name Resolver
       DNS Names
         Should resolve localhost properly:
     Error: Timeout of 4000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.

  2) Name Resolver
       DNS Names
         Should default to port 443:
     Error: Timeout of 4000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
```

However, those tests were failing from the moment I pulled down the repo; I assumed it was a quirk of the local environment (or some minor dependency mismatch somewhere).

Reply:
Yes, I meant to remove the 5 byte header from the output of that function, and the other code you referenced that removes it again.

Don't worry about those test failures. I see the same failures on WSL; it seems to be a quirk of how Windows or WSL resolves `localhost`.

Reply:
Sounds good, pushed a commit that simplifies the removal of the "prefix".